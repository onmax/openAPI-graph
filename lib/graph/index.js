"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAPIGraphsManager = exports.OpenAPIGraphsBuilder = exports.OpenAPIGraph = void 0;
var OpenAPIGraph_1 = require("./OpenAPIGraph");
Object.defineProperty(exports, "OpenAPIGraph", { enumerable: true, get: function () { return OpenAPIGraph_1.OpenAPIGraph; } });
var OpenAPIGraphsBuilder_1 = require("./OpenAPIGraphsBuilder");
Object.defineProperty(exports, "OpenAPIGraphsBuilder", { enumerable: true, get: function () { return OpenAPIGraphsBuilder_1.OpenAPIGraphsBuilder; } });
var OpenAPIGraphsManager_1 = require("./OpenAPIGraphsManager");
Object.defineProperty(exports, "OpenAPIGraphsManager", { enumerable: true, get: function () { return OpenAPIGraphsManager_1.OpenAPIGraphsManager; } });
