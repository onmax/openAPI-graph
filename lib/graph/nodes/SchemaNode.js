"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaNode = void 0;
const Node_1 = require("./Node");
class SchemaNode extends Node_1.Node {
    constructor(name, content) {
        super(name);
        this.content = content;
    }
}
exports.SchemaNode = SchemaNode;
