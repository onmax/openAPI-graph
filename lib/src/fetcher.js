"use strict";
/**
 * The main goal is to find all openAPI specification given a path.
 * Therefore, it will loop over all files and creating a structured object
 * as a result
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.testables = exports.fetcher = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const openapi_1 = require("./openapi");
async function fetcher(path) {
    // Converts path to absolute
    path = path_1.resolve(path);
    const pathExists = fs_1.existsSync(path);
    if (!pathExists) {
        throw new Error('The given path does not exist in your system');
    }
    else {
        return await loadsSwaggerFiles(path);
    }
}
exports.fetcher = fetcher;
/**
 * Finds all the valid OpenApi specifications given a directory and returns a
 * list of all the specification contents
 *
 * @param projectPath of the project where the specifications are suppose to be
 * @returns a list of OpenAPIContent which has the path, openAPI version
 *          and the content for every specification found inside of the given path
 */
async function loadsSwaggerFiles(projectPath) {
    const projectContent = fs_1.readdirSync(projectPath);
    if (projectContent === undefined || projectContent.length === 0) {
        return [];
    }
    const paths = projectContent
        // Find all yml and yaml files
        .filter((p) => p.match(/.*\.(yml|yaml)/gi))
        .map(p => `${projectPath}/${p}`);
    return await openapi_1.getOpenApisContent(paths);
}
// Just for testing reasons https://stackoverflow.com/a/54116079
exports.testables = {
    loadsSwaggerFiles,
};
