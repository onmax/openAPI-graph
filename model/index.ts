export { OpenAPIContent } from './OpenAPIContent';
export { Nodes } from './Nodes';
export { Edges } from './Edges';
export { EdgesRefDict } from './EdgesRefDict';
export { RefType, RefTo } from './RefType';
