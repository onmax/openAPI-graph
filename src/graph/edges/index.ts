export { Edge } from './Edge';
export { RefEdge } from './RefEdge';
