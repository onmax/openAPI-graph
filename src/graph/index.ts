export { OpenAPIGraph } from './OpenAPIGraph';
export { OpenAPIGraphsBuilder } from './builder/OpenAPIGraphsBuilder';
export { OpenAPIGraphs } from './OpenAPIGraphs';
