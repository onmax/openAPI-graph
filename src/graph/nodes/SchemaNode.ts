import { OpenAPIV3 } from 'openapi-types';
import { SchemaNodeConstructor, SchemaNodeInterface } from 'openapi-graph-types';
import { Node } from './Node';

export const SchemaNode: SchemaNodeConstructor = class SchemaNodeImpl extends Node implements SchemaNodeInterface {
  content!: OpenAPIV3.SchemaObject;

  constructor(name: string, content: OpenAPIV3.SchemaObject) {
    super(name);
    this.content = content;
  }
}
