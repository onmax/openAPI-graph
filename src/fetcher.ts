/**
 * The main goal is to find all openAPI specification given a path.
 * Therefore, it will loop over all files and creating a structured object
 * as a result
 *
 */

import { existsSync, readdirSync } from 'fs';
import { resolve } from 'path';
import { OpenAPIContent } from '../model';
import { getOpenApisContent } from './openapi';

export async function fetcher(path: string): Promise<OpenAPIContent[]> {
  // Converts path to absolute
  path = resolve(path);
  const pathExists = existsSync(path);

  if (!pathExists) {
    throw new Error('The given path does not exist in your system');
  } else {
    return await loadsSwaggerFiles(path);
  }
}

/**
 * Finds all the valid OpenApi specifications given a directory and returns a
 * list of all the specification contents
 *
 * @param projectPath of the project where the specifications are suppose to be
 * @returns a list of OpenAPIContent which has the path, openAPI version
 *          and the content for every specification found inside of the given path
 */
async function loadsSwaggerFiles(projectPath: string): Promise<OpenAPIContent[]> {
  const projectContent = await getFiles(resolve(projectPath));
  if (projectContent === undefined || projectContent.length === 0) {
    return [];
  }

  return await getOpenApisContent(projectContent);
}

async function getFiles(fromPath = './', paths: string[] = []): Promise<string[]> {
  const entries = await readdirSync(fromPath, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.isDirectory()) {
      paths.push(...(await getFiles(`${fromPath}/${entry.name}/`)));
    } else if (entry.name.match(/.*\.(yml|yaml|json)/gi)) {
      paths.push(resolve(`${fromPath}/${entry.name}`));
    }
  }
  return paths;
}

// Just for testing reasons https://stackoverflow.com/a/54116079
export const testables = {
  loadsSwaggerFiles,
  getFiles,
};
