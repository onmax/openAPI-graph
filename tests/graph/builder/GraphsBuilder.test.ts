import { RefType } from '../../../model';
import { fetcher } from '../../../src/fetcher';
import { OpenAPIGraph, OpenAPIGraphsBuilder } from '../../../src/graph';
import { RefEdge } from '../../../src/graph/edges';
import { SchemaNode } from '../../../src/graph/nodes/SchemaNode';

test('Creates a graph from the petstore specification', async () => {
    const petstoreApis = await fetcher("tests/resources/petstore");
    const petstoreBuilder = new OpenAPIGraphsBuilder(petstoreApis);
    expect(petstoreBuilder.graphs).toHaveLength(1);
    expect(petstoreBuilder.graphs[0]).toMatchSnapshot({
        path: expect.stringContaining('tests/resources/petstore'),
        edges: expect.any(Object),
        nodes: expect.any(Object),
    });
    const graph = petstoreBuilder.graphs[0]
    const schemasNames: string[] = Object.values(graph.getSchemaNodes()).map(n => n.name).sort();
    const expected = ["Pet", "Pets", "Error", "SchemaNotBeingUsed"].sort()
    expect(schemasNames).toEqual(expected)

    const refSchemas: RefEdge[] = Object.values(graph.getSchemaRefEdges());
    const expectedSchemas = ['Pets', 'Error', 'Pet'].sort()
    expect(refSchemas.map(n => n.ref).sort()).toStrictEqual(expectedSchemas.map(s => `#/components/schemas/${s}`))
    expect(refSchemas.map(n => n.tokenName).sort()).toStrictEqual(expectedSchemas)
});

test('Creates a graph from the social-network specification', async () => {
    const socialNetworkApis = await fetcher("tests/resources/social-network");
    const socialNetworkBuilder = new OpenAPIGraphsBuilder(socialNetworkApis);
    expect(socialNetworkBuilder.graphs).toHaveLength(3);

    const socialNetworkGraph = socialNetworkBuilder.graphs.find(g => g.path.endsWith('social-network.yaml')) as OpenAPIGraph
    const socialNetworkUsersGraph = socialNetworkBuilder.graphs.find(g => g.path.endsWith('users.yaml')) as OpenAPIGraph
    const socialNetworkPostsGraph = socialNetworkBuilder.graphs.find(g => g.path.endsWith('posts.yaml')) as OpenAPIGraph

    expect(socialNetworkGraph).toMatchSnapshot({
        path: expect.stringContaining('tests/resources/social-network/social-network.yaml'),
    });
    expect(socialNetworkUsersGraph).toMatchSnapshot({
        path: expect.stringContaining('tests/resources/social-network/users'),
    });
    expect(socialNetworkPostsGraph).toMatchSnapshot({
        path: expect.stringContaining('tests/resources/social-network/posts'),
    });

    expect(Object.values(socialNetworkGraph.edges.ref.schemaRef)).toHaveLength(4);
    expect(Object.values(socialNetworkUsersGraph.edges.ref.schemaRef)).toHaveLength(2);
    expect(Object.values(socialNetworkPostsGraph.edges.ref.schemaRef)).toHaveLength(1);

    // Test social-network.yaml
    const expected = ["Username", "Name", "Email", "User", "Post"].sort()
    const schemasSocialNetworkNames: string[] = Object.values(socialNetworkGraph.getSchemaNodes()).map(n => n.name).sort();
    expect(schemasSocialNetworkNames).toEqual(expected)
    const refSocialNetworkNames: RefEdge[] = Object.values(socialNetworkGraph.getSchemaRefEdges());
    const expectedSchemas = ['Username', 'Name', 'Email', 'User'].sort()
    expect(refSocialNetworkNames.map(n => n.ref).sort()).toStrictEqual(expectedSchemas.map(s => `#/components/schemas/${s}`))
    expect(refSocialNetworkNames.map(n => n.tokenName).sort()).toStrictEqual(expectedSchemas)

    // // Test users.yaml
    const schemasUsersNames: string[] = Object.values(socialNetworkUsersGraph.getSchemaNodes()).map(n => n.name).sort();
    expect(schemasUsersNames).toStrictEqual([])
    Object.values(socialNetworkUsersGraph.getSchemaRefEdges()).map(edge => {
        expect(edge.type).toStrictEqual(RefType.Remote)
        expect(edge.child).toBeInstanceOf(SchemaNode)
    })
    expect(Object.values(socialNetworkUsersGraph.getSchemaRefEdges()).map(edge => edge.tokenName).sort()).toStrictEqual(['User', 'Username'].sort())

    // // Test posts.yaml
    const schemasPostsNames: string[] = Object.values(socialNetworkPostsGraph.getSchemaNodes()).map(n => n.name).sort();
    expect(schemasPostsNames).toStrictEqual([])
    Object.values(socialNetworkPostsGraph.getSchemaRefEdges()).map(edge => {
        expect(edge).toBeInstanceOf(RefEdge)
        expect(edge.type).toStrictEqual(RefType.Remote)
        expect(edge.child).toBeInstanceOf(SchemaNode)
    })
    expect(Object.values(socialNetworkPostsGraph.getSchemaRefEdges()).map(edge => edge.tokenName).sort()).toStrictEqual(['Post'])
});