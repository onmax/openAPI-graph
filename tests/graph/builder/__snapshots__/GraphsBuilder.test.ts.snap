// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates a graph from the petstore specification 1`] = `
Object {
  "edges": Any<Object>,
  "nodes": Any<Object>,
  "path": StringContaining "tests/resources/petstore",
}
`;

exports[`Creates a graph from the social-network specification 1`] = `
Object {
  "edges": Object {
    "ref": Object {
      "schemaRef": Object {
        "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml#/components/schemas/Email": RefEdgeImpl {
          "absolutePath": "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml",
          "child": SchemaNodeImpl {
            "content": Object {
              "format": "email",
              "type": "string",
            },
            "name": "Email",
          },
          "ref": "#/components/schemas/Email",
          "tokenName": "Email",
          "type": "local",
        },
        "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml#/components/schemas/Name": RefEdgeImpl {
          "absolutePath": "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml",
          "child": SchemaNodeImpl {
            "content": Object {
              "type": "string",
            },
            "name": "Name",
          },
          "ref": "#/components/schemas/Name",
          "tokenName": "Name",
          "type": "local",
        },
        "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml#/components/schemas/User": RefEdgeImpl {
          "absolutePath": "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml",
          "child": SchemaNodeImpl {
            "content": Object {
              "properties": Object {
                "email": Object {
                  "$ref": "#/components/schemas/Email",
                },
                "name": Object {
                  "$ref": "#/components/schemas/Name",
                },
                "username": Object {
                  "$ref": "#/components/schemas/Username",
                },
              },
              "required": Array [
                "username",
                "name",
                "email",
              ],
              "type": "object",
            },
            "name": "User",
          },
          "ref": "#/components/schemas/User",
          "tokenName": "User",
          "type": "local",
        },
        "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml#/components/schemas/Username": RefEdgeImpl {
          "absolutePath": "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml",
          "child": SchemaNodeImpl {
            "content": Object {
              "type": "string",
            },
            "name": "Username",
          },
          "ref": "#/components/schemas/Username",
          "tokenName": "Username",
          "type": "local",
        },
      },
    },
  },
  "nodes": Object {
    "schemas": Object {
      "Email": SchemaNodeImpl {
        "content": Object {
          "format": "email",
          "type": "string",
        },
        "name": "Email",
      },
      "Name": SchemaNodeImpl {
        "content": Object {
          "type": "string",
        },
        "name": "Name",
      },
      "Post": SchemaNodeImpl {
        "content": Object {
          "properties": Object {
            "author": Object {
              "$ref": "#/components/schemas/User",
            },
            "content": Object {
              "type": "string",
            },
            "date": Object {
              "type": "string",
            },
          },
          "required": Array [
            "author",
            "date",
            "content",
          ],
          "type": "object",
        },
        "name": "Post",
      },
      "User": SchemaNodeImpl {
        "content": Object {
          "properties": Object {
            "email": Object {
              "$ref": "#/components/schemas/Email",
            },
            "name": Object {
              "$ref": "#/components/schemas/Name",
            },
            "username": Object {
              "$ref": "#/components/schemas/Username",
            },
          },
          "required": Array [
            "username",
            "name",
            "email",
          ],
          "type": "object",
        },
        "name": "User",
      },
      "Username": SchemaNodeImpl {
        "content": Object {
          "type": "string",
        },
        "name": "Username",
      },
    },
  },
  "path": StringContaining "tests/resources/social-network/social-network.yaml",
}
`;

exports[`Creates a graph from the social-network specification 2`] = `
Object {
  "edges": Object {
    "ref": Object {
      "schemaRef": Object {
        "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml#/components/schemas/User": RefEdgeImpl {
          "absolutePath": "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml",
          "child": SchemaNodeImpl {
            "content": Object {
              "properties": Object {
                "email": Object {
                  "$ref": "#/components/schemas/Email",
                },
                "name": Object {
                  "$ref": "#/components/schemas/Name",
                },
                "username": Object {
                  "$ref": "#/components/schemas/Username",
                },
              },
              "required": Array [
                "username",
                "name",
                "email",
              ],
              "type": "object",
            },
            "name": "User",
          },
          "ref": "../social-network.yaml#/components/schemas/User",
          "tokenName": "User",
          "type": "remote",
        },
        "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml#/components/schemas/Username": RefEdgeImpl {
          "absolutePath": "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml",
          "child": SchemaNodeImpl {
            "content": Object {
              "type": "string",
            },
            "name": "Username",
          },
          "ref": "../social-network.yaml#/components/schemas/Username",
          "tokenName": "Username",
          "type": "remote",
        },
      },
    },
  },
  "nodes": Object {
    "schemas": Object {},
  },
  "path": StringContaining "tests/resources/social-network/users",
}
`;

exports[`Creates a graph from the social-network specification 3`] = `
Object {
  "edges": Object {
    "ref": Object {
      "schemaRef": Object {
        "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml#/components/schemas/Post": RefEdgeImpl {
          "absolutePath": "/home/max/openapi-graph-core/tests/resources/social-network/social-network.yaml",
          "child": SchemaNodeImpl {
            "content": Object {
              "properties": Object {
                "author": Object {
                  "$ref": "#/components/schemas/User",
                },
                "content": Object {
                  "type": "string",
                },
                "date": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "author",
                "date",
                "content",
              ],
              "type": "object",
            },
            "name": "Post",
          },
          "ref": "../social-network.yaml#/components/schemas/Post",
          "tokenName": "Post",
          "type": "remote",
        },
      },
    },
  },
  "nodes": Object {
    "schemas": Object {},
  },
  "path": StringContaining "tests/resources/social-network/posts",
}
`;
